rudder-core/src/main/scala/com/normation/rudder/services/policies/write/BuildBundleSequence.scala

```diff
@@ -0,0 +1,373 @@
+  ////////// Implementation details //////////
+  ////////////////////////////////////////////
+
+
+
+  /*
+   * For each techniques, build:
+   * - the promiser
+   * - the list of bundle included,
+   */
+  def buildTechniqueBundles(nodeId: NodeId)(t3: (Technique, List[BundleOrder], PolicyMode)): TechniqueBundles = {
+    // naming things to make them cleare
+    val technique  = t3._1
+    val promiser   = Promiser(t3._2.map(_.value).mkString("/"))
+    val policyMode = t3._3
+
+    // We need to remove zero-length bundle name from the bundlesequence (like, if there is no ncf bundles to call)
+    // to avoid having two successives commas in the bundlesequence
+    val techniqueBundles = technique.bundlesequence.flatMap { bundle =>
+      if(bundle.name.trim.size > 0) {
+        Some(bundle)
+      } else {
+        logger.warn(s"Technique '${technique.id}' used in node '${nodeId.value}' contains some bundle with empty name, which is forbidden and so they are ignored in the final bundle sequence")
+        None
+      }
+    }.toList
+
+    TechniqueBundles(promiser, Nil, techniqueBundles, Nil, technique.isSystem, technique.providesExpectedReports, policyMode)
+  }
+
+  /*
+   * Some Techniques don't have any bundle (at least common).
+   * We don't want to include these technique in the bundle sequence,
+   * obviously
+   */
+  implicit final class NoBundleTechnique(bundles: List[TechniqueBundles]) {
+    def removeEmptyBundle: List[TechniqueBundles] = bundles.filterNot(_.main.isEmpty)
+  }
+
+  /*
+   * The logic that is in charge to add ncf reporting information
+   * by calling the correct bundle when appropriate
+   */
+  implicit final class NcfReporting(bundles: List[TechniqueBundles]) {
+    //now, for each technique that provided reports (i.e: an ncf technique), we must add the
+    //NCF_REPORT_DEFINITION_BUNDLE_NAME just before the other bundle of the technique
+    def addNcfReporting: List[TechniqueBundles] = bundles.map { tb =>
+      //we assume that the bundle name to use as suffix of NCF_REPORT_DEFINITION_BUNDLE_NAME
+      // is the first of the provided bundle sequence for that technique
+      if(tb.providesExpectedReports) {
+        val newMain = tb.main match {
+          case Nil => Nil
+          case firstBundle :: tail => (Bundle(s"""current_technique_report_info("${firstBundle.name}")""")
+                                       :: firstBundle :: tail)
+        }
+        tb.copy(main = newMain)
+      } else {
+        tb
+      }
+    }
+  }
+
+  /*
+   * Each bundle must be preceded by the correct "set_dry_mode" mode,
+   * and we always end with a "set_dry_mode("false")".
+   * Also, promiser must be differents for all items so that cfengine
+   * doesn't try to avoid to do the set, so we are using the technique
+   * promiser.
+   */
+  implicit final class DryRunManagement(bundles: List[TechniqueBundles]) {
+    def addDryRunManagement: List[TechniqueBundles] = bundles match {
```

Comment:
> if we have a global dry mode enabled, shouldn't we let the set_dry_mode("true") ?

Well, we discussed it internally, but: if we do that, it means that we want to alter the behavior for bundle outside user bundle, so necesserally system bundle (as in "used by Rudder to make Rudder works, not by the user to make his infra works") - by definition. The consensus seems to have been that "dry mode" is different to "rudder agent does nothing on that node", and "dry mode" was still implying that the rudder inner tools need to work as expected (and so, error should be corrected).